include/sys/dmu_objset.h

```diff
@@ -120,13 +133,12 @@ struct objset {
 
 	/* Protected by os_obj_lock */
 	kmutex_t os_obj_lock;
+	struct dnode_list os_dnode_list[OS_DNODE_LISTS];
+	int os_dirty[TXG_SIZE];
 	uint64_t os_obj_next;
 
 	/* Protected by os_lock */
 	kmutex_t os_lock;
-	list_t os_dirty_dnodes[TXG_SIZE];
-	list_t os_free_dnodes[TXG_SIZE];
-	list_t os_dnodes;
```

Comment:
You may be able to simplify this by using the `multilist_t` functionality provided in `multilist.c`.  Each `multilist_t` is constructed from N sub-lists and an index function can be provided to control which sublist an entry is inserted in.  You can also lock a specific sub-list's if needed using the `multilist_sublist_lock()` function.